src/wp-includes/fonts/class-wp-font-library.php

```diff
@@ -0,0 +1,136 @@
+<?php
+/**
+ * Font Library class.
+ *
+ * This file contains the Font Library class definition.
+ *
+ * @package    WordPress
+ * @subpackage Font Library
+ * @since      6.4.0
+ */
+
+/**
+ * Font Library class.
+ *
+ * @since 6.4.0
+ */
+class WP_Font_Library {
+
+	/**
+	 * Fonts mime types allowed for each file type.
+	 * Each file type can have multiple mime types depending on the PHP version.
+	 * Currently wp_check_filetype_and_ext() only allows one mime type per file extension.
+	 */
+	const PHP_7_TTF_MIME_TYPE = PHP_VERSION_ID >= 70300 ? 'application/font-sfnt' : 'application/x-font-ttf';
+
+	const ALLOWED_FONT_MIME_TYPES = array(
+		'otf'   => 'font/otf',
+		'ttf'   => PHP_VERSION_ID >= 70400 ? 'font/sfnt' : self::PHP_7_TTF_MIME_TYPE,
+		'woff'  => PHP_VERSION_ID >= 80100 ? 'font/woff' : 'application/font-woff',
+		'woff2' => PHP_VERSION_ID >= 80100 ? 'font/woff2' : 'application/font-woff2',
+	);
+
+	/**
+	 * Font collections.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @var array
+	 */
+	private static $collections = array();
+
+	/**
+	 * Register a new font collection.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param array $config Font collection config options.
+	 *                      See {@see wp_register_font_collection()} for the supported fields.
+	 * @return WP_Font_Collection|WP_Error A font collection is it was registered successfully and a WP_Error otherwise.
+	 */
+	public static function register_font_collection( $config ) {
+		$new_collection = new WP_Font_Collection( $config );
+
+		if ( isset( self::$collections[ $config['id'] ] ) ) {
+			return new WP_Error( 'font_collection_registration_error', 'Font collection already registered.' );
+		}
+
+		self::$collections[ $config['id'] ] = $new_collection;
+		return $new_collection;
+	}
+
+	/**
+	 * Gets all the font collections available.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return array List of font collections.
+	 */
+	public static function get_font_collections() {
+		return self::$collections;
+	}
+
+	/**
+	 * Gets a font collection.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @param string $id Font collection id.
+	 * @return array List of font collections.
+	 */
+	public static function get_font_collection( $id ) {
+		if ( array_key_exists( $id, self::$collections ) ) {
+			return self::$collections[ $id ];
+		}
+		return new WP_Error( 'font_collection_not_found', 'Font collection not found.' );
+	}
+
+	/**
+	 * Gets the upload directory for fonts.
+	 *
+	 * @since 6.4.0
+	 *
+	 * @return string Path of the upload directory for fonts.
+	 */
+	public static function get_fonts_dir() {
+		return path_join( WP_CONTENT_DIR, 'fonts' );
+	}
```

Comment:
Instead of storing the file here, why not in `wp-content/uploads/font`. Many hosts do not allow writing to wp-content and make a special rule for `wp-content/uploads`.

Reply:
https://github.com/WordPress/gutenberg/issues/53965

Reply:
@matiasbenedetto Have you tested with different hosts and with plugins like s3 uploads?

Reply:
Does it make sense to make the location a configuration option?

Reply:
I think it is. `WP_CONTENT` is configurable: https://developer.wordpress.org/apis/wp-config-php/#moving-wp-content-folder
So the font directory is, too, because it's relative to `WP_CONTENT`.

Reply:
Also, like a special case for wp-content/uploads so might they for wp-content/fonts (if users installing fonts is important to them)

Reply:
I don't see it as an essential feature, and it isn't necessary for the release of 6.4 beta 2 (the target of this PR) either. Could it be discussed in a Gutenberg issue, maybe?

Reply:
>Instead of storing the file here, why not in `wp-content/uploads/font`. Many hosts do not allow writing to wp-content and make a special rule for `wp-content/uploads`.

Where to store the fonts was discussed during the library's development by folks like @azaozz @costdev @mtias @matiasbenedetto me and others. To help bring the context from those discussions to here, I'm copying links and comments.

@azaozz discussed locating the fonts in `wp-content/fonts` [here in the original GB PR](https://github.com/WordPress/gutenberg/pull/52704#discussion_r1271167455). Coping his comment here for readability:

>Thinking more about this, the `/fonts` directory will be part of WP, just like `/plugins`, `/themes`, `/uploads`, etc. right? Then it probably makes the most sense to just add it (in git/svn) under `/wp-content` like the other directories that hold various content/assets.
>
>In that case the `define_fonts_directory()` and `create_fonts_directory()` would probably be better moved out of the class as they make sense only for the Gutenberg plugin. The code from them can perhaps be added to a "Gutenberg helper function" that runs on `plugins_loaded` and ensures the constant and the directory exist. Perhaps something like this (untested):
```php
if ( ! defined( 'WP_FONTS_DIR' ) ) {
    define( 'WP_FONTS_DIR', WP_CONTENT_DIR . '/fonts' );
}

if ( ! is_dir( WP_FONTS_DIR ) ) {
    wp_mkdir_p( WP_FONTS_DIR );
}
```

A follow-up change was made as requested by @mtias to move the fonts back to `wp-content/fonts`, see https://github.com/WordPress/gutenberg/issues/53965.

Reply:
Coping some of the discussion from Make/Core slack to wire everything together:

@spacedmonkey [asked](https://wordpress.slack.com/archives/C02RQBWTW/p1696258580604899):
>Regarding the localtion of the font files. Should there be a way to configure where they are stored, like uploads.

@azaozz [replied](https://wordpress.slack.com/archives/C02RQBWTW/p1696270309785459?thread_ts=1696258580.604899&cid=C02RQBWTW):
>Thinking this is not a good idea. Would you want to allow someone to move their `/plugins` directory under `/uploads` , or perhaps under `/themes`  ? :slightly_smiling_face:

Reply:
@matiasbenedetto Assume this is now possible via [the `font_dir` filter?](https://github.com/WordPress/gutenberg/blob/280084ed188c6d18db4300c53f2ec17ca50f2749/lib/compat/wordpress-6.5/fonts/fonts.php#L210)?

Reply:
> @matiasbenedetto Assume this is now possible via [the font_dir filter?](https://github.com/WordPress/gutenberg/blob/280084ed188c6d18db4300c53f2ec17ca50f2749/lib/compat/wordpress-6.5/fonts/fonts.php#L210)?

That's correct.